subsys/usb/device_next/usbd_msg.c

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdio.h>
+#include <zephyr/kernel.h>
+#include <zephyr/init.h>
+#include <zephyr/usb/usbd.h>
+
+#include "usbd_device.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(usbd_msg, CONFIG_USBD_LOG_LEVEL);
+
+static void msg_work_handler(struct k_work *work);
+static K_WORK_DEFINE(msg_work, msg_work_handler);
+K_FIFO_DEFINE(msg_queue);
+
+struct usbd_msg_pkt {
+	sys_snode_t node;
+	struct usbd_contex *ctx;
+	struct usbd_msg msg;
+};
+
+K_MEM_SLAB_DEFINE_STATIC(usbd_msg_slab, sizeof(struct usbd_msg_pkt),
+			 CONFIG_USBD_MAX_MESSAGES, sizeof(void *));
+
+static ALWAYS_INLINE int submit_new_work(struct usbd_msg_pkt *const m_pkt)
+{
+	k_fifo_put(&msg_queue, m_pkt);
+	return k_work_submit(&msg_work);
+}
+
+static inline struct usbd_msg_pkt *usbd_alloc_msg(struct usbd_contex *const ctx)
+{
+	struct usbd_msg_pkt *m_pkt;
+
+	if (k_mem_slab_alloc(&usbd_msg_slab, (void **)&m_pkt, K_NO_WAIT)) {
+		LOG_ERR("Failed to allocate message memory");
+		return NULL;
+	}
+
+	memset(m_pkt, 0, sizeof(struct usbd_msg_pkt));
+	m_pkt->ctx = ctx;
+
+	return m_pkt;
+}
+
+static void msg_work_handler(struct k_work *work)
+{
+	struct usbd_msg_pkt *m_pkt;
+	struct usbd_contex *ctx;
+	int err;
+
+	m_pkt = k_fifo_peek_head(&msg_queue);
+	if (m_pkt == NULL) {
+		return;
+	}
+
+	ctx = m_pkt->ctx;
+	err = ctx->msg_cb(&m_pkt->msg);
+	if (err == -EBUSY || err == -EAGAIN) {
+		LOG_DBG("Message callback is busy (%d)", err);
+		(void)k_work_submit(work);
+		return;
+	}
+
+	m_pkt = k_fifo_get(&msg_queue, K_NO_WAIT);
+	if (m_pkt != NULL) {
+		k_mem_slab_free(&usbd_msg_slab, (void *)m_pkt);
+	}
+
+	if (!k_fifo_is_empty(&msg_queue)) {
+		(void)k_work_submit(work);
+	}
+}
+
+int usbd_msg_register_cb(struct usbd_contex *const uds_ctx,
+			 const usbd_msg_cb_t cb)
+{
+	int ret = 0;
+
+	usbd_device_lock(uds_ctx);
+
+	if (uds_ctx->msg_cb != NULL) {
+		ret = -EALREADY;
+		goto register_cb_exit;
+	}
+
+	uds_ctx->msg_cb = cb;
+
+register_cb_exit:
+	usbd_device_unlock(uds_ctx);
+
+	return ret;
+}
+
+int usbd_msg_pub_simple(struct usbd_contex *const ctx,
+			const enum usbd_msg_type type, const int status)
+{
+	struct usbd_msg_pkt *m_pkt;
+	int err;
+
+	if (ctx->msg_cb == NULL) {
+		return -ENOTSUP;
```

Comment:
Propably a personal preference, but I wouldn't consider this to be an error condition. I would just `return 0;` here.

Reply:
reworked, error can not be forwarded anyway.